pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java

```diff
@@ -295,7 +314,16 @@ public PgConnection(HostSpec[] hostSpecs,
     initObjectTypes(info);
 
     if (PGProperty.LOG_UNCLOSED_CONNECTIONS.getBoolean(info)) {
-      openStackTrace = new Throwable("Connection was created at this point:");
+      Throwable stackTrace = new Throwable("Connection was created at this point:");
+      Object leakMarker = new Object();
+      this.leakMarker = leakMarker;
+      this.leakTraceHandle =
+          ConnectionLeakTrackerHolder.INSTANCE.register(leakMarker, stackTrace);
+
+      ConnectionLeakTrackerHolder.INSTANCE.processReferences((leak) ->
+          LOGGER.log(Level.WARNING, GT.tr("Finalizing a Connection that was never closed:"),
+              leak.stackTrace)
```

Comment:
shouldn't we always be doing this, not just when logging is enabled?

Reply:
I believe Java closes network resources anyway, so the only extra we should have is to display "leak detected" and clear pending timers.

Reply:
What about the shared cancelTimer?
How certain are you that Java will close the socket in similar timeframe as identifying phantom reference onto reference queue? The comments on the existing finalize method seem to indicate Java’s default behavior for de-referenced sockets was not sufficient.